### PR TITLE
Move `cocotb.runner` and `cocotb.config`

### LINKED ISSUE #3731: Move `cocotb.runner` and `cocotb.config` to new package
The runner.py code being in the cocotb package requires `cocotb/__init__.py` to be evaluated, which evaluates the whole damn project causing slow startup times. This could be alleviated by moving `cocotb.runner` and `cocotb.config` into a new package (perhaps `cocotb_tools`). Re-export stubs would be added with deprecation warning.

### PR DESCRIPTION
**Pull request by ktbarrett**
Closes #3731.

### CODE REVIEW

**Comment by ktbarrett** [comment]
For some reason I can't replicate the error locally even when setting `PYTHONWARNINGS`, so imma just gonna keep pushing.

**Comment by ktbarrett** [comment]
Apparently `warnings.simplefilter` simply does not work across imports.

**Comment by ktbarrett** [comment]
I haven't the slightest clue... this stuff is busted.

**Review by cmarqu — APPROVED** [review]

**Comment by ktbarrett** [comment]
I guess the lineno in the error isn't the lineno you should use to suppress it. The lineno did appear to be wrong, but locally the wrong lineno still suppressed the message? IDK... I just hope there isn't any other `UserWarning`s coming out of `cocotb_tools.runner`.

File: src/cocotb_tools/runner.py [hunk 1]
```diff
@@ -1027,6 +1034,8 @@ def _build_command(self) -> List[Command]:
         # TODO: set "--debug" if self.verbose
         # TODO: support "--always"
 
+        import cocotb
+
         verilator_cpp = str(
             Path(cocotb.__file__).parent
             / "share"
```
Comment (ktbarrett) [inline src/cocotb_tools/runner.py]: Refactor this to use `cocotb_tools.config.share_dir`

File: src/cocotb_tools/config.py [hunk 2]
```diff
@@ -47,21 +47,27 @@
 
 import find_libpython
 
-import cocotb
+import cocotb_tools
 
-__all__ = ["share_dir", "makefiles_dir", "libs_dir"]
+base_tools_dir = Path(cocotb_tools.__file__).parent.resolve()
+base_cocotb_dir = base_tools_dir.parent.joinpath("cocotb").resolve()
```
Comment (ktbarrett) [inline src/cocotb_tools/config.py:53]: Add check to see if this exists and if not then `import cocotb` and use that instead.

File: src/cocotb_tools/makefiles/Makefile.inc [hunk 3]
```diff
@@ -47,7 +47,7 @@ COCOTB_PY_DIR := $(shell cocotb-config --prefix)
 
 # Directory containing all support files required to build cocotb-based
 # simulations: Makefile fragments, and the simulator libraries.
-COCOTB_SHARE_DIR := $(COCOTB_PY_DIR)/cocotb/share
+COCOTB_SHARE_DIR := $(shell cocotb-config --share)
```
Comment (ktbarrett) [inline src/cocotb_tools/makefiles/Makefile.inc:50]: This doesn't need to be done in this commit. OR take the opportunity to deprecate a remove some variables that are equivalent to cocotb-config calls.

File: docs/source/building.rst [hunk 4]
```diff
@@ -411,36 +411,3 @@ and
```
Comment (marlonjames) [inline docs/source/building.rst:411]: Edit path to new config.py on line 41 above
```
  If you edit the following sections, please also update the "helpmsg" text in src/cocotb/config.py
```

File: src/cocotb_tools/runner.py [hunk 5]
```diff
@@ -0,0 +1,1266 @@
...
+        if self.hdl_toplevel is None:
+            raise ValueError(
+                f"{type(self).__qualname__}: Simulator requires the hdl_toplevel parameter to be specified"
+            )
+
+        # TODO: set "--debug" if self.verbose
+        # TODO: support "--always"
+
+        verilator_cpp = str(
+            cocotb_tools.config.share_dir / "lib" / "verilator" / "verilator.cpp"
```
[the remark this answers is not in the document]
Comment (ktbarrett) [reply src/cocotb_tools/runner.py:1038]: Mmmmm. Should we change this to a Path object?

**Comment by ktbarrett** [comment]
Changing to Path object was a PITA, but it's a bit better this way IMO.

File: pyproject.toml [hunk 6]
```diff
@@ -127,7 +127,7 @@ source = [
 
 [tool.coverage.report]
 omit = [
-    "*/cocotb/config.py",
```
[the remark this answers is not in the document]
Comment (ktbarrett) [reply pyproject.toml:131]: Well we have 100% coverage on that file.

**Review by marlonjames — APPROVED** [review]

**Pull request merged**